### ISSUE TITLE
Order is lost for more than 10 routes

### ISSUE THREAD

**Issue by terhechte** [metosin/compojure-api#138]
Hey,
it seems that the ordering of routes is only preserved if there are less than or equal to 10 routes. Above that, the ordering will switch to random. I've played around with this with my current spec of 16 routes, and the order comes back as I reach the critical treshold of 10, and disappears as soon as I uncomment any of the commented out routes - not important which one.

Is there something I can do to enforce ordering, maybe structure things into a vector? I had a look at the codebase but couldn't find anything that helped.

Thanks,
Benedikt

**Comment by ikitommi** [comment]
Using a vector seems to be a good idea, the order will persist over AOT compilation too. Seems that the route order gets lost on the ring-swagger side. Will check this after the summer vacations.

https://github.com/metosin/compojure-api/blob/master/test/compojure/api/swagger_ordering_test.clj

**Comment by terhechte** [comment]
Thanks! by using a vector you mean:

``` Clojure
(defapi api
    (context "/a" []
      [(GET* "/1" [] identity)
      (GET* "/2" [] identity)
      ...]
```

?

**Comment by ikitommi** [comment]
Hi. No changes to user facing apis. We can already collect the routes top-down in order, but the calculated route-tree is internally stored in a map (path->handler), which kills the order for all hash-maps. So a internal change. Mostly works already ;)

**Comment by ikitommi** [comment]
fixed this on the ring-swagger side (https://github.com/metosin/ring-swagger/commit/6b82c64b8410c5aa6ad45f8eb860d344870bfe92), will be fixed on c-api side in the next minor release.

**Comment by Deraen** [comment]
This doesn't work yet. Looks like it non-trivial to start using ordered-map in Compojure-api. Added tests here: https://github.com/metosin/compojure-api/compare/keep-paths-ordering?expand=1

1. `extract-routes` loses the order when doing deep-merge: https://github.com/metosin/compojure-api/blob/master/src/compojure/api/swagger.clj#L234
2. Looks like using syntax quote is problematic with ordered map: https://github.com/metosin/compojure-api/blob/master/src/compojure/api/routes.clj#L76-L77. It tries to write code with `#ordered/map` tag or something.

**Comment by ikitommi** [comment]
I have a working code for this. I think. Will do PR today.

**Comment by ikitommi** [comment]
ok. this turned out to be really non-trivial - many clojure.core -fns don't seem to respect the (special) map types. I guess we should start using vector of routes with ring-swagger.

**Comment by ikitommi** [comment]
in my branch, used internally vectors, reformed them as maps when reading. OrderedMaps don't work with AOT, so read it back into an ArrayMap (https://github.com/metosin/compojure-api/blob/669f5e2a56bafe205e2387c9305cfcf5e45097be/src/compojure/api/routes.clj#L70). Which will revert back to HashMap under the hoods as an implementation detail with `assoc` etc.

**Comment by ikitommi** [comment]
seems to work now - https://github.com/metosin/compojure-api/compare/ordering?expand=1&w=1

```clojure
(ns examples.ordered
  (:require [compojure.api.sweet :refer :all]
            [ring.util.http-response :refer :all]))

(defroutes* more-ordered-routes
  (GET* "/6" [] (ok))
  (GET* "/7" [] (ok))
  (GET* "/8" [] (ok)))

(defroutes* ordered-routes
  (context* "/ordered" []
    :tags ["ordered"]
    (context* "/a" []
      (GET* "/1" [] (ok))
      (GET* "/2" [] (ok))
      (GET* "/3" [] (ok))
      (context* "/b" []
        (GET* "/4" [] (ok))
        (GET* "/5" [] (ok)))
      (context* "/c" []
        more-ordered-routes
        (GET* "/9" [] (ok))
        (GET* "/10" [] (ok))))))
```

<img width="940" alt="ordering" src="https://cloud.githubusercontent.com/assets/567532/9569584/54163228-4f78-11e5-8daf-fb6c9296d276.png">

**Comment by ikitommi** [comment]
should be fixed now, with `0.23.1`.
